docs/proposals/30-apiserver-proxy.md

```diff
@@ -0,0 +1,326 @@
+4. Graduate the `RemoveAPIServerProxyLegacyPort` feature gate to GA and lock to `True`.
+   - Remove the `APIServerProxyUsesHTTPProxy` constraint in the `Shoot` status.
+
+After these steps, the proxy protocol is no longer used and the related network infrastructure (including the `proxy` port 8443) is removed.
+The remaining steps are related to unifying the HTTP proxy infrastructure:
+
+1. Introduce the unified HTTP proxy network infrastructure as described in [this section](#unifying-the-http-proxy-infrastructure).
+   - Reconfigure the API server proxy and shoot VPN client to connect to the unified port.
+   - Report a new constraint `UsesUnifiedHTTPProxyPort` in the `Shoot` status once both components have been reconfigured to use the new port.
+2. Introduce a new feature gate `RemoveHTTPProxyLegacyPort` to gardenlet.
+   - If enabled, gardenlet removes the old HTTP proxy network infrastructure on the seed side (ingress gateway port `tls-tunnel`, `Gateway`, `EnvoyFilter`, etc.)
+   - The gardenlet only allows activating the feature gate if all shoot clusters on its seed cluster report that the API server proxy and shoot VPN client have been reconfigured.
+3. Graduate the `RemoveHTTPProxyLegacyPort` feature gate to GA and lock to `True`.
+   - Remove the `UsesUnifiedHTTPProxyPort` constraint in the `Shoot` status.
+
+After these steps, the legacy VPN network infrastructure on the seed side (including the `tls-tunnel` port 8132) is removed. Instead, the unifying infrastructure using the `http-proxy` port 8443 is used.
+
+In this plan, we need the feature gates for guarding the removal of one of the ingress gateway ports because we cannot be sure that a given Gardener version has been able to successfully reconfigure the related components in all shoot clusters.
+I.e., we cannot assume that the port in question is no longer used, because the shoot reconciliation flow might not have reached the relevant tasks.
+This might be because the shoot is marked to be ignored or an `Extension` with lifecycle `BeforeKubeAPIServer` cannot be reconciled successfully.
+The feature gate allows the Gardener operator to delay the removal of the ingress gateway port if it's still used in the environment.
+With this, we ensure the Gardener operator is not blocked in upgrading to a newer version.
+
+The constraints are used as a mechanism for reporting which shoot clusters still use the old configuration of API server proxy and shoot VPN client.
+This helps the Gardener operator to determine easily whether the corresponding feature gate can be enabled.
+Additionally, gardenlet can prevent misconfiguration by iterating the relevant shoot constraints.
+
+## Alternatives
+
+We are not aware of any other alternative solution to address this issue, as requiring a transparent LoadBalancer for Gardener is no solution.
+
+## Appendix
```

Comment:
I'm tempted to drop the diagrams.
I feel like it's hard to visualize the architecture accurately while still making it helpful for understanding the topic.
I would hope that the precise descriptions in the document are good enough to understand the proposal.

WDYT?

Reply:
Generally, visualisations can support describing the architecture. Therefore, I am usually all for having diagrams.
However, I can see why it is hard in this case to find the right visualisation. Having good diagrams is great, but getting there might be time-consuming.
Unless you have a good idea how to visualise it, feel free to drop them.

TLDR: I am fine either way.

Reply:
I gave this a shot and reworked the diagrams to visualize the layered connections.
The diagrams are not perfectly accurate (layer numbers, destination IPs, header values, etc.), but they should be good enough to make the proposal easier to understand.

Looking forward to your feedback :)